### PR TITLE
[SPARK-42832][SQL] Remove repartition if it is the child of LocalLimit

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR enhances `CollapseRepartition` to remove repartition if it is the child of `LocalLimit`. Because its output is determined by the number of partitions and the expressions of the Repartition. Therefore, it is feasible to remove Repartition except for repartition by nondeterministic expressions, because users may expect to randomly take data.
For example:
```sql
SELECT /*+ REBALANCE */ * FROM t WHERE id > 1 LIMIT 5;
```

Before this PR:
```
== Optimized Logical Plan ==
GlobalLimit 5
+- LocalLimit 5
   +- RebalancePartitions
      +- Filter (isnotnull(id#0L) AND (id#0L > 1))
         +- Relation spark_catalog.default.t[id#0L] parquet
```

After this PR:
```
== Optimized Logical Plan ==
GlobalLimit 5
+- LocalLimit 5
   +- Filter (isnotnull(id#0L) AND (id#0L > 1))
      +- Relation spark_catalog.default.t[id#0L] parquet
```

Note that we don't remove repartition if it looks like the user might want to take data randomly. For example:
```sql
SELECT /*+ REPARTITION(3) */ * FROM t WHERE id > 1 LIMIT 5;
SELECT * FROM t WHERE id > 1 DISTRIBUTE BY random() LIMIT 5;
```

### Why are the changes needed?

Reduce shuffle to improve query performance. The use case is that we add a repartition to improve the parallelism on a JDBC table:

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
cc @sunchao

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -1213,6 +1213,12 @@ object CollapseRepartition extends Rule[LogicalPlan] {
     // child.
     case r @ RebalancePartitions(_, child: RebalancePartitions, _) =>
       r.withNewChildren(child.children)
+    // Case 5: When a LocalLimit has a child of Repartition we can remove the Repartition.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1216]: why we can remove it? can we put more explanation in the PR description?

**Comment by VindhyaG** [comment]
Can you please explain more on scenarios when rebalancepartitions becomes child of locallimit? i tried  
SELECT  * FROM t WHERE id > 1 LIMIT 5; with spark 2.4.4 version and rebalancepartitions is not part of the plan.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1216]: OK

**Comment by VindhyaG** [comment]
> > Can you please explain more on scenarios when rebalancepartitions becomes child of locallimit? i tried SELECT * FROM t WHERE id > 1 LIMIT 5; with spark 2.4.4 version and rebalancepartitions is not part of the plan.
> 
> Spark 2.4.4 does not have `RebalancePartitions`. Please see: https://issues.apache.org/jira/browse/SPARK-35786
Got it.thanks!

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM from my side.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.5.

**Comment by cloud-fan** [comment]
I don't quite get the rationale. For `SELECT /*+ REBALANCE */ * FROM t WHERE id > 1 LIMIT 5;`, the user explicitly requires to do a rebalance before limit, why do we remove it? It's a physical hint and we should respect it.

**Comment by wangyum** [comment]
rebalance before limit feels meaningless.
If the user really wants to take n rows from each partition, the SQL should look like this:
```sql
SELECT *
FROM   (SELECT *,
               Row_number() OVER(partition BY pmod(Hash(id, 42), 200) ORDER BY id) AS rn
        FROM   tbl)
WHERE  rn <= 5;
```

**Comment by cloud-fan** [comment]
The optimizer can remove meanless logical operators as it won't change the query result and it's usually more efficient to remove operators. But for physical APIs like rebalance and repartition, the users explicitly ask to take control of performance tuning and we should trust them. Besides, I don't think this is an always-beneficial optimization, what if the limit is very large? We should let the user make the decision: they can remove the rebalance hint if they believe it makes the perf worse.  What do you think? @wangyum @dongjoon-hyun

**Comment by wangyum** [comment]
Could we push the `LocalLimit` through `RebalancePartitions`?
```
== Optimized Logical Plan ==
GlobalLimit 5
+- LocalLimit 5
   +- RebalancePartitions
      +- LocalLimit 5
         +- Filter (isnotnull(id#0L) AND (id#0L > 1))
            +- Relation spark_catalog.default.tbl[id#0L] parquet
```

**Comment by cloud-fan** [comment]
I don't think so. Local limit is per partition and rebalance changes partitioning.

**Comment by cloud-fan** [comment]
oh wait, I thought the hint applies as the last operator. `SELECT /*+ REBALANCE */ * FROM t WHERE id > 1 LIMIT 5;` should rebalance be the root node for this query?

**Comment by dongjoon-hyun** [comment]
`HINT` is a part of `SELECT` clause.
https://github.com/apache/spark/blob/a3d9e0ae0f95a55766078da5d0bf0f74f3c3cfc3/sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4#L530-L532

That's the reason why `LIMIT` is the root of the query.
```
scala> sql("SELECT /*+ REBALANCE */ * FROM t WHERE id > 1 LIMIT 5").explain(true)
== Parsed Logical Plan ==
'GlobalLimit 5
+- 'LocalLimit 5
   +- 'UnresolvedHint REBALANCE
      +- 'Project [*]
         +- 'Filter ('id > 1)
            +- 'UnresolvedRelation [t], [], false
```

**Comment by wangyum** [comment]
If such queries cannot be optimized, the performance of such queries will be very poor. We use a partition to fetch data from MySQL, and increase its parallelism for downstream computing after fetching the data:

```sql
CREATE VIEW full_query_log
AS
SELECT h.* FROM query_log_hdfs h
UNION ALL
SELECT /*+ REBALANCE */ q.*, DATE(start) FROM query_log_mysql q;

SELECT * FROM full_query_log limit 5;
```

**Comment by cloud-fan** [comment]
OK now I got the use case. At the time when we add the rebalance hint, we don't know what the final query is. Shall we make this optimization a bit more conservative to match both global and local limit? I still think it's risky to remove rebalance below local limit.

**Comment by dongjoon-hyun** [comment]
+1 for @cloud-fan 's suggestion.

**Comment by dongjoon-hyun** [comment]
Could you make a follow-up, @wangyum ?

**Comment by wangyum** [comment]
How about reverting this PR directly? Otherwise, we need to change `LimitPushDown`, which is not elegant:
```scala
case Limit(exp, u: Union) if u.children.exists(_.isInstanceOf[HasPartitionExpressions]) =>
  val newChildren = u.children.map {
    case r: RebalancePartitions =>
      maybePushLocalLimit(exp, r.child)
    case r: RepartitionByExpression
        if r.partitionExpressions.nonEmpty && r.partitionExpressions.forall(_.deterministic) =>
      maybePushLocalLimit(exp, r.child)
    case o =>
      maybePushLocalLimit(exp, o)
  }
  Limit(exp, u.copy(children = newChildren))
```

**Comment by cloud-fan** [comment]
why do we need to change `LimitPushDown`? We can't push limit through repartition.

**Comment by wangyum** [comment]
It is meaningless to only optimize this case, because the user can manually remove the hint:
```sql
SELECT /*+ REBALANCE */ * FROM tbl LIMIT 5;
```

I'd like to optimize this case, since the user cannot manually remove the hint, because the view may be maintained by others:
```sql
CREATE VIEW full_query_log
AS
SELECT h.* FROM query_log_hdfs h
UNION ALL
SELECT /*+ REBALANCE */ q.*, DATE(start) FROM query_log_mysql q;

SELECT * FROM full_query_log limit 5;
```

**Comment by wangyum** [comment]
Reverted this commit: https://github.com/apache/spark/commit/58facc16536cb2784aace38d6cca77bd055a3053.

**Comment by dongjoon-hyun** [comment]
Thank you for the decision, @wangyum .

**Pull request closed without merging**